File: src/main/java/LeetCode/LeetCode53.java
package LeetCode;

/**
 * @author jingxinwu
 * @date 2021-11-16 11:21 下午
 */
public class LeetCode53 {

    public int maxSubArray(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        //dp[i] 以i为结尾的子数组
        int[] dp = new int[nums.length];
        int res = nums[0];
        dp[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            dp[i] = nums[i] + (Math.max(dp[i - 1], 0));
            res = Math.max(res, dp[i]);
        }
        return res;
    }

}

File: src/main/go/leetcode/lists/hot100/LeetCode53.go
package hot100

//[-2,1,-3,4,-1,2,1,-5,4]
func maxSubArray(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	res := nums[0]
	for i := 1; i < len(nums); i++ {
		dp[i] = Max(dp[i-1], 0) + nums[i]
		res = Max(dp[i], res)
	}
	return res
}
